### PR TITLE
[RLlib] Prioritized multi-agent episode replay buffer.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

This PR adds prioritized sampling for multi-agent setups. It implements `"independent"` sampling by holding sum- and min-segments for each module and updates them accordingly by given TD-errors (more generally by `priorities`).

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -180,6 +180,7 @@ java/**/pom.xml
 # python virtual env
 venv
 .venv
+.venv*
```
Comment (sven1977) [inline .gitignore:183]:
```suggestion
```

File: rllib/algorithms/sac/sac.py [hunk 2]
```diff
@@ -355,6 +355,7 @@ def validate(self) -> None:
             "EpisodeReplayBuffer",
             "PrioritizedEpisodeReplayBuffer",
             "MultiAgentEpisodeReplayBuffer",
+            "MultiAgentPrioritizedEpisodeReplayBuffer",
```
Comment (sven1977) [inline rllib/algorithms/sac/sac.py:358]: Yeah, this is getting very long :D

File: rllib/algorithms/sac/torch/sac_torch_learner.py [hunk 3]
```diff
@@ -314,14 +310,23 @@ def compute_gradients(
             for component in (
                 ["qf", "policy", "alpha"] + ["qf_twin"] if config.twin_q else []
             ):
-                self.metrics.peek((module_id, component + "_loss")).backward(
+                # Get the optimizer for the current component and module.
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/sac_torch_learner.py:313]: Nice! Did this help with learning performance a little?

File: rllib/algorithms/sac/torch/sac_torch_learner.py [hunk 4]
```diff
@@ -314,14 +310,23 @@ def compute_gradients(
...
+                self.metrics.peek(module_id, component + "_loss").backward(
                     retain_graph=True
                 )
+                # Store the gradients for the component and module.
+                # TODO (simon): Check another time the graph for overlapping
+                # gradients.
                 grads.update(
                     {
-                        pid: p.grad
+                        pid: p.grad.clone()
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/sac_torch_learner.py:327]: Can you explain why we need to `clone` this?

Comment (sven1977) [reply rllib/algorithms/sac/torch/sac_torch_learner.py:327]: Add a comment with the explanation.

Maybe this is the same problem I have in torch DreamerV3 ...

File: rllib/tuned_examples/dqn/multi_agent_cartpole_dqn.py [hunk 5]
```diff
@@ -14,7 +14,7 @@
 # Use `parser` to add your own custom command line options to this script
 # and (if needed) use their values to set up `config` below.
 args = parser.parse_args()
-
+args.num_agents = 2
```
Comment (sven1977) [inline rllib/tuned_examples/dqn/multi_agent_cartpole_dqn.py:17]: nit: better: parser.set_default(num_agents=2)

then inside the __main__:
```
assert args.num_agents > 0, "The `--num-agents` arg must be > 0 for this script to work."
```

File: rllib/tuned_examples/dqn/multi_agent_cartpole_dqn.py [hunk 6]
```diff
@@ -63,4 +65,6 @@
 if __name__ == "__main__":
     from ray.rllib.utils.test_utils import run_rllib_example_script_experiment
 
+    args.local_mode = True
```
Comment (sven1977) [inline rllib/tuned_examples/dqn/multi_agent_cartpole_dqn.py:68]: nit: no local mode

On enable_new_api_stack:
Either "bake" it into the config below (config.api_stack(...)) 
OR force - through an assert - the user to provide the `--enable-new-api-stack`.

File: rllib/tuned_examples/sac/multi_agent_pendulum_sac.py [hunk 7]
```diff
@@ -35,10 +44,12 @@
         train_batch_size_per_learner=256,
         target_network_update_freq=1,
         replay_buffer_config={
-            "type": "MultiAgentEpisodeReplayBuffer",
+            "type": "MultiAgentPrioritizedEpisodeReplayBuffer",
```
Comment (sven1977) [inline rllib/tuned_examples/sac/multi_agent_pendulum_sac.py:47]: Noice!

File: rllib/tuned_examples/sac/multi_agent_pendulum_sac.py [hunk 8]
```diff
@@ -35,10 +44,12 @@
...
             "capacity": 100000,
+            "alpha": 1.0,
+            "beta": 0.0,
         },
-        num_steps_sampled_before_learning_starts=1024,
+        num_steps_sampled_before_learning_starts=256,
```
Comment (sven1977) [inline rllib/tuned_examples/sac/multi_agent_pendulum_sac.py:52]: 👍

File: rllib/utils/replay_buffers/episode_replay_buffer.py [hunk 9]
```diff
@@ -350,6 +354,15 @@ def get_added_timesteps(self) -> int:
 
     @override(ReplayBufferInterface)
     def get_state(self) -> Dict[str, Any]:
+        """Gets a pickable state of the buffer.
```
Comment (sven1977) [inline rllib/utils/replay_buffers/episode_replay_buffer.py:357]: Thanks for adding these docstrings ...

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 10]
```diff
@@ -0,0 +1,924 @@
+import copy
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:1]: Solution for the long names bloat:

* `multi_agent` and `single_agent` (clarity is good)
* `replay_buffer` -> `buffer`
* `prioritized` -> `prio` ?
* `episode` -> `` <- all buffers should be based on episodes (but this is probably only possible in the long run after all batch based buffers are gone).

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 11]
```diff
@@ -0,0 +1,924 @@
...
+    corresponding (importance sampling) weight for the transition.
+
+    After sampling priorities can be updated (for the last sampled episode list) with
+    `self.update_priorities`. This method assigns the new priorities automatically to
+    the last sampled timesteps. Note, this implies that sampling timesteps and updating
+    their corresponding priorities needs to alternate (e.g. sampling several times and
+    then updating the priorities would not work because the buffer caches the last
+    sampled timestep indices).
+
+    .. testcode::
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:60]: Nice, thanks for adding this test.

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 12]
```diff
@@ -0,0 +1,924 @@
...
+
+        # Initialize the first episode entries.
+        eps = MultiAgentEpisode()
+        obs, infos = env.reset()
+        eps.add_env_reset(observations=obs, infos=infos)
+
+        # Sample 10,000 env timesteps.
+        for i in range(num_timesteps):
+            # If terminated we create a new episode.
+            if terminateds["__all__"] or truncateds["__all__"]:
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:89]: nit: `eps.is_done` is simpler.

File: .gitignore (hunk 1 above)
Comment (simonsays1980) [reply .gitignore:183]: This would try to stage all virtual environments that are not of name `.venv`  exactly.

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 13]
```diff
@@ -0,0 +1,924 @@
...
+                obs,
+                actions,
+                rewards,
+                infos,
+                terminateds=terminateds,
+                truncateds=truncateds
+            )
+
+        # Add the last (truncated) episode to the list of episodes.
+        if not terminateds["__all__"] or truncateds["__all__"]:
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:116]: same: `if not eps.is_done:`

File: rllib/algorithms/sac/sac.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/algorithms/sac/sac.py:358]: In the future we can remove all `Episode` in the names.

File: rllib/algorithms/sac/torch/sac_torch_learner.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/algorithms/sac/torch/sac_torch_learner.py:313]: Yeah  it was somehow a combination with some hyperparameters.

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 14]
```diff
@@ -0,0 +1,924 @@
...
+    ) -> None:
+        """Adds incoming episodes to the replay buffer.
+
+        Note, if the incoming episodes' time steps cause the buffer to overflow,
+        older episodes are evicted. Because episodes usually come in chunks and
+        not complete, this could lead to edge cases (e.g. with very small capacity
+        or very long episode length) where the first part of an episode is evicted
+        while the next part just comes in.
+        In such cases, we evict the complete episode, including the new chunk,
+        unless the episode is the last one in the buffer. In the latter case the
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:212]: "last one" -> "only one"

File: rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py [hunk 15]
```diff
@@ -41,28 +41,96 @@ class MultiAgentEpisodeReplayBuffer(EpisodeReplayBuffer):
...
-    Each batch contains - in addition to the data tuples presented above - two further
-    columns, namely `n_steps` and `weigths`. The former holds the `n_step` used for each
-    row in the batch and the latter the (default) weight of `1.0` for each row in the
-    batch. The weight is used for weighted loss calculations in the training process.
+    Each episode contains - in addition to the data tuples presented above - two further
+    elements in its `extra_model_outputs`, namely `n_steps` and `weights`. The former
+    holds the `n_step` used for the sampled timesteps in the episode and the latter the
+    corresponding (importance sampling) weight for the transition.
+
+    .. testcode::
```
Comment (sven1977) [inline rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py:49]: Nice!

File: rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py [hunk 16]
```diff
@@ -460,7 +582,7 @@ def _sample_independent(
         include_infos: bool,
         include_extra_model_outputs: bool,
         modules_to_sample: Optional[Set[ModuleID]],
-    ) -> SampleBatchType:
+    ) -> List["SingleAgentEpisode"]:
```
Comment (sven1977) [inline rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py:585]: nice!

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 17]
```diff
@@ -0,0 +1,924 @@
...
+            while B < batch_size_B:
+                # First, draw a random sample from Uniform(0, sum over all weights).
+                # Note, transitions with higher weight get sampled more often (as
+                # more random draws fall into larger intervals).
+                module_random_sum = (
+                    self.rng.random() * self._module_to_sum_segment[module_id].sum()
+                )
+                # Get the highest index in the sum-tree for which the sum is
+                # smaller or equal the random sum sample.
+                # Note, we sample `o_(t + n_step)` as this is the state that
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:803]: Should we better say: "We sample the transition `o_(t)` to `o_(t + n_step)` as this ..."

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 18]
```diff
@@ -0,0 +1,924 @@
...
+                sa_episode = ma_episode.agent_episodes[agent_id]
+
+                # If we use random n-step sampling, draw the n-step for this item.
+                if random_n_step:
+                    actual_n_step = int(self.rng.integers(n_step[0], n_step[1]))
+                # If we cannnot make the n-step, we resample.
+                if sa_episode_ts + actual_n_step > len(sa_episode):
+                    continue
+                # Note, this will be the reward after executing action
+                # `a_(episode_ts)`. For `n_step>1` this will be the sum of
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:844]: -> "discounted sum"

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py [hunk 19]
```diff
@@ -0,0 +1,924 @@
...
+                # all rewards that were collected over the last n steps.
+                sa_raw_rewards = sa_episode.get_rewards(
```
Comment (sven1977) [inline rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:846]: perfect!

Comment (sven1977) [reply rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:846]: I'm always worried about the DreamerV3 logic, which is different from all other algos wrt to sampling rewards from the buffer. The first reward in an episode is always 0.0 (it's basically a "reset" reward, similar to the "reset" zeroed-out-action).

Just need to be careful when we subclass the Dreamer's single-agent episode buffer.

File: rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_episode_replay_buffer.py [hunk 20]
```diff
@@ -0,0 +1,335 @@
...
+        self.assertTrue(
+            {eps.id_ for eps in buffer.episodes}
+            == {"3", "4", "5", "6", "7", "8", "9", "G"}
+        )
+        for module_id in buffer.get_module_ids():
+            # self.assertTrue(buffer.get_num_episodes(module_id) == 8)
+            self.assertTrue(buffer.get_num_timesteps(module_id) == 91)
+            self.assertTrue(buffer.get_added_timesteps(module_id) == 424)
+
+    # def test_buffer_independent_sample_logic(self):
```
Comment (sven1977) [inline rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_episode_replay_buffer.py:145]: Should we include this test (functionality is done in the buffer class, no)?

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py (hunk 10 above)
Comment (simonsays1980) [reply rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py:1]: I fully agree with the comments. At this moment we have arrived at the longest name possible and withe the old stack deprecated we can release the `Episode`  in the names. 

I will rename the `ma_...` to ` multi_agent_...`

File: rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_episode_replay_buffer.py (hunk 20 above)
Comment (simonsays1980) [reply rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_episode_replay_buffer.py:145]: Yes we could include this fucntionality (`get/set_state`).

In a new PR we have to cleanup the sampling tests for "independent" sampling and the `add`  method.

File: rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py (hunk 17 above)
Comment (simonsays1980) [reply rllib/utils/replay_buffers/ma_prioritized_episode_replay_buffer.py]: I changed this comment completely as it is out-dated: "Note, in contrast to Schaul et al. (2018) (who sample `o_(t + n_step)`, Algorithm 1) we sample `o_t`."

File: rllib/algorithms/sac/torch/sac_torch_learner.py (hunk 4 above)
Comment (simonsays1980) [reply rllib/algorithms/sac/torch/sac_torch_learner.py:327]: This merely a safeguard to avoid the gradients stored for certain parameters being inadvertedly changed by the optimizer in further operations.

**Review by sven1977 — APPROVED** [review]
LGTM now! Thanks for all the fixes.
